### DISCUSSION TITLE
Best practice to maintain requirements for different environments?

### DISCUSSION THREAD

**Discussion by huxuan** [pdm-project/pdm#2058] [Q&A] [answered]
In certain cases, we may want to maintain requirements for different environments. For example, different Python versions, different operation system (Linux/Mac/Windows), different architecture (x86/arm64), or even different third-party packages (PyTorch 1.X, PyTorch 2.X).

My current solution is to maintain a list of minimal requirements with different constraints. The idea comes from airflow. It would be great if we can make it with more powerful dependency management tools, but seems pipenv, poetry and pdm only support kinds of hash lock to restrict the environment rather than support multiple environments simultaneously.

**Comment by sanmai-NL** [comment, marked as answer]
You can use multiple lockfiles using `--lockfile`. You can also make the lockfile less generic: https://pdm.fming.dev/latest/usage/dependency/#cross-platform-lockfile.

**Reply by sanmai-NL** [reply]
See also: https://github.com/pdm-project/pdm/issues/1787

**Reply by huxuan** [reply]
Thanks for the answer and sorry for the late reply. I will try it later.
